design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,350 @@
+ ```
+
+While in the Azure Authenticator the Resource Restrictions are always stored in
+the host annotations, in the Kubernetes Authenticator the Resource Restrictions
+can be stored either in the host's ID or in its annotations. 
+For example:
+
+ ```yaml
+- !host
+  id: test-app-service-account
+  annotations:
+    authn-k8s/namespace: some-namespace
+    authn-k8s/service-account: some-service-account
+
+- !host
+  id: some-namespace/service_account/some-service-account
+ ```
+
+In the example above, both hosts have the same Resource Restrictions. One stores 
+it in its annotation and the other in its ID. However, their Conjur identity is
+not the same.
+
+## Issue Description
+
+When we add a new authenticator to Conjur, we need to add the mechanism
+of validating the host's Resource Restrictions. Currently we have no generic
+way to do it and because of that it is less straight-forward to add a new
+authenticator. As can be seen in the implementations for the
+[Azure Authenticator](/app/domain/authentication/authn_azure/validate_resource_restrictions.rb)
+and 
+[Kubernetes Authenticator](/app/domain/authentication/authn_k8s/validate_resource_restrictions.rb), 
+there is a lot of code duplication with minor changes. 
+
+## Solution
+
+For the sake of simplicity, from here and throughout the document, Resource
+Restrictions will be defined only in the host's annotations, and not in the ID. 
+In future authenticators we will not develop an option to define it in the ID
+and it is available in the K8s Authenticator only for backwards compatibility.
+
+### Defining Resource Restrictions Validation
+
+The Resource Restrictions validation can be split into 3 parts:
+ 1. Extract the Resource Restrictions from the host
+ 2. Validate the Resource Restrictions are defined correctly
+ 3. Validate the Resource Restrictions match the authentication request
+
+Let's break down each part to better understand them.
+
+#### 1. Extract the Resource Restrictions from the host
+
+This step goes to the host in the policy and extracts all its annotations
+that starts with `authenticator_name` or `authenticator_name/service_id`.
+
+It should return a hash of all the restrictions, without their prefix.
+
+The class gets the host's annotations and the authenticator's service-id, 
+and then builds a `resource_restrictions` hash from the annotations. 
+
+In this case the general class will be defined as follows:
+- input
+  - authenticator name
+  - authenticator service-id
+  - account
+  - host name
+- output
+  - `resource_restrictions` hash.
+  
+
+So this class signature is:
+
+```ruby
+ExtractResourceRestrictions = CommandClass.new(
+  dependencies: {
+    role_class:                  ::Role,
+    resource_class:              ::Resource,
+    logger:                      Rails.logger
+  },
+  inputs:       %i(authenticator_name service_id host_name account)
+) do
+
+  def call
+    ensure_role_exists
+    init_role_annotations
```

Comment:
`init_role_annotations` what does this call do? does it initialize some role annotation object?

Reply:
or fetch the annotations for the defined host/user?

Reply:
It gets **all** the role annotations defined for the `host_name` in the policy.

Reply:
so many `fetch_all_role_annotations` would be better from a naming standpoint